pkg/storage/unified/resource/kv.go

```diff
@@ -0,0 +1,213 @@
+package resource
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"iter"
+	"time"
+
+	badger "github.com/dgraph-io/badger/v4"
+)
+
+var ErrNotFound = errors.New("key not found")
+
+type SortOrder int
+
+const (
+	SortOrderUndefined SortOrder = iota
+	SortOrderAsc
+	SortOrderDesc
+)
+
+type ListOptions struct {
+	Sort     SortOrder // sort order of the results. Default is SortOrderAsc.
+	StartKey string    // lower bound of the range, included in the results
+	EndKey   string    // upper bound of the range, excluded from the results
+	Limit    int64     // maximum number of results to return. 0 means no limit.
+}
+
+// KVReader represents a key-value pair with streaming value access
+type KVReader struct {
+	Key    string
+	Reader io.Reader
+}
+
+type KV interface {
+	// Keys returns all the keys in the store
+	Keys(ctx context.Context, section string, opt ListOptions) iter.Seq2[string, error]
+
+	// Get retrieves a key-value pair from the store
+	Get(ctx context.Context, section string, key string) (KVReader, error)
+
+	// Save a new value from an io.Reader
+	Save(ctx context.Context, section string, key string, value io.Reader) error
+
+	// Delete a value
+	Delete(ctx context.Context, section string, key string) error
+
+	// UnixTimestamp returns the current time in seconds since Epoch.
+	// This is used to ensure the server and client are not too far apart in time.
+	UnixTimestamp(ctx context.Context) (int64, error)
+}
+
+// Reference implementation of the KV interface using BadgerDB
+// This is only used for testing purposes, and will not work HA
+type badgerKV struct {
+	db *badger.DB
+}
+
+func NewBadgerKV(db *badger.DB) *badgerKV {
+	return &badgerKV{
+		db: db,
+	}
+}
+
+func (k *badgerKV) Get(ctx context.Context, section string, key string) (KVReader, error) {
+	txn := k.db.NewTransaction(false)
+	defer txn.Discard()
+
+	if section == "" {
+		return KVReader{}, fmt.Errorf("section is required")
+	}
+
+	key = section + "/" + key
+
+	item, err := txn.Get([]byte(key))
+	if err != nil {
+		if errors.Is(err, badger.ErrKeyNotFound) {
+			return KVReader{}, ErrNotFound
+		}
+		return KVReader{}, err
+	}
+
+	out := KVReader{
+		Key: string(item.Key())[len(section)+1:],
+	}
+
+	// Get the value and create a reader from it
+	var value []byte
+	err = item.Value(func(val []byte) error {
+		value = make([]byte, len(val))
+		copy(value, val)
+		return nil
+	})
+	if err != nil {
+		return KVReader{}, err
+	}
+
+	out.Reader = bytes.NewReader(value)
```

Comment:
When using `io.ReadCloser`:

```suggestion
	out.Reader = io.NopCloser(bytes.NewReader(value))
```